### DISCUSSION TITLE
Train yolo detection models with auto labeling

### DISCUSSION THREAD

**Discussion by eyasu6464** [ultralytics/ultralytics#23313] [Show and tell]
I wanted to share a workflow/tool I’ve been using to generate labeled data for training YOLO detectors, especially when the target concepts are too specific or not covered by standard datasets.

The idea is simple: instead of manually annotating everything from scratch, use an open-vocabulary, language-conditioned detector to auto-label real images from free-form prompts (not a fixed class list), then export those boxes in YOLO format and fine-tune a standard YOLO model on top.

This has been useful for cases like:
very specific 

> object definitions (e.g. semantic or attribute-based concepts)
> bootstrapping datasets to check if a concept is learnable before full annotation
> reducing the initial labeling cost while still training on real-world data

Important clarification:
This is not meant to fully replace proper human annotation. Real data + cleaned labels still matter. The goal is to speed up dataset creation so YOLO training starts from something better than zero.
I built a small tool/API around this idea that takes an image + natural-language prompt and returns bounding boxes that can be directly used as YOLO labels. It’s been helpful in my own training pipelines, so I’m sharing it here for feedback.

I’d be interested to hear:
how 

> others here bootstrap datasets for custom detectors
> limitations you see with auto-labeling in YOLO workflows
> edge cases you think this approach will fail on

Tool : https://www.useful-ai-tools.com/tools/detect-anything
Full data sampling → auto-labeling → training workflow Colab: https://colab.research.google.com/github/useful-ai-tools/detect-anything/blob/main/notebooks/yolo_dataset_builder_and_trainer.ipynb

**Comment by UltralyticsAssistant** [comment]
👋 Hello @eyasu6464, thank you for sharing your workflow with the Ultralytics community 🚀🧰 — auto-labeling and bootstrapping datasets is a topic a lot of folks are exploring right now, and it’s great to see real-world tooling being discussed here.

This is an automated response 🤖, and an Ultralytics engineer will also assist soon.

A couple quick notes to help the discussion stay actionable:
-If you’re looking for feedback on how this fits into YOLO training pipelines, it helps to include a small end-to-end example of the exported YOLO labels plus a short training log excerpt (even just a few iterations) so others can comment on quality + failure modes.
-If you’ve run into any issues integrating the generated labels into Ultralytics training, please share details (and if it’s a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can reproduce locally) 🧪.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered 📚.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community 💬.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
This is a solid “pseudo-label → human clean-up → fine-tune” workflow, and you can do the same entirely inside Ultralytics using prompted open-vocabulary models like [YOLOE](https://docs.ultralytics.com/models/yoloe/) (text prompts via `set_classes()` / `classes=...`), then export boxes with `save_txt=True` and fine-tune a standard detector; the main limitations we see in practice are prompt ambiguity/synonyms (“red bag” vs “suitcase”), attribute grounding failures (colors, materials, “damaged”), and domain shift (lighting/cameras) causing systematic false positives/loose boxes, so it helps to start with a high `conf`, manually audit a small stratified sample, and iterate (and after any label edits, delete `labels/*.cache` so training re-reads updated labels).
